---
title: "UK Biobank Imputed Genotype"
author: "Xiaoran Tong"
output: html_document
---

# preperation

Make sure  the main phenotype has  been downloaded from UKB  (see [dwn.Rmd][] or
[dwn.html][]),  and  variables  that   describe  genotyping  quality  have  been
extracted  (see section  __Genotyping Process  and Quality__  in [phe.Rmd][]  or
[phe.html][]).

Make sure that the imputed genotype is have been download from UKB or accessable
from the institution (i.e., MSU).

The downloaded files are in Oxford  BGEN format, together with meta information,
kept under __{s}/ukb/imp/000__.

**Note**: {s} denotes  the root of scratch  space, which is big  enough to store
 and process  imputed genotype, but  will be purged  after a specific  period of
 time (44 days for MSU HPCC).

To ease future  work, from within __{p}/dat__ create links  to imputed genotypes
under __{s}/ukb/imp__.  Make sure the data is visible to other group members.

```{sh, eval=TRUE}
p=$GRP/ukb; cd $p				# project directory - UKB related
ln -sf $SCR/ukb/imp dat/imp		# create link: dat/imp -> imputed
ls -gGhL dat/imp/000/*			# inspect the link
```

-------------------------------------------------------------------------------

[dwn.Rmd]: https://github.com/xiaoran831213/ukb/blob/master/src/dwn.Rmd "download from UKB, Rmarkdown"
[dwn.html]: https://htmlpreview.github.io/?https://github.com/xiaoran831213/ukb/blob/master/doc/dwn.html "download from UKB, HTML"
[phe.Rmd]: https://github.com/xiaoran831213/ukb/blob/master/src/phe.Rmd "extract phenotype, Rmarkdown"
[phe.html]: https://htmlpreview.github.io/?https://github.com/xiaoran831213/ukb/blob/master/doc/phe.html "extract phenotype, HTML"

# Extraction

Extracting  The  entire  imputation cost  large  disk  space and a  long Running
jobs, use the following critera to scale down the data:

  * information score $\ge$ 0.3;
  * MAF $\ge$ 0.005$;

Write the result in PLINK2 __PGEN__ format.

```{sh, eval=FALSE}
p=$GRP/ukb; cd $p				# project directory
w=dat/imp; cd $w				# working directory
s=000							# downloads
d=001; mkdir -p $d				# well-imputed variants
chr=({0..22} X Y XY MT)			# chromosome: character
num=({00..26})					# chromosome: double digits 
for i in {1..26}; do
	n=${num[$i]}; c=${chr[$i]}; o=$d/$n # chr-num, chr-name, and output
    # source exists?
    if [ ! -e $s/$n.bgen ]; then echo "echo miss $s/$n"; continue; fi
	# target exists?
	if [ -e $o.pgen -a -e $o.psam -a -e $o.pvar ]; then echo "echo skip $o"; continue; fi
    
    # list well-imputed variants 
    [ -e $o.mfi ] || awk <$s/$n.mfi      '$6>=.005 && $8>=.3' >$o.mfi
    [ -e $o.pos ] || awk <$o.mfi -v c=$i '{print c":"$3}'     >$o.pos
    [ -e $o.vid ] || awk <$o.mfi -v c=$i '{print $2}'         >$o.vid

    # bgen -> pgen
    echo -n "time plink2 --bgen 000/$n.bgen ref-first --sample 000/$n.sample "
	echo    "--extract $n.vid --make-pgen --out $n --memory 4096 --threads 1"
done | tee tmp.cmd
hpcwp tmp.cmd -d$d -q1 --wtm 12 -m8 --ln $s --log none; rm tmp.cmd
```
The extraction make took up to 8 hours for longer chromosomes.

-------------------------------------------------------------------------------

# Quality Control

## Variants
Retain variants satisfying the follow criterias:

  * HWE test p-value > 1e-6
    - for autosomes and PAR (1-22 and 25), use all samples.
    - for X chromosome without PAR (23), use females only;
	- for Y chromosome, use males only (not in UKB).
  * call rate $\ge$ 95%
  * non-duplicates (not enforced);
  * bi-allelic SNP only (not enforeced for dosage data)

**Note**: even for imputed  genotype, call rate will not reach  100% due to hard
call threshold specified in the previous step (extration).

PLINK2 automatically takes special care of HWE test on X/Y Chromosome.

```{sh, eval=FALSE}
cd $GRP/ukb/dat/imp				# working directory
s=001							# imputation extracted
d=002; mkdir -p $d				# QC

## find duplicated variant
for f in $s/*.pvar; do
	## sorted variant id
	tail -n+2 $f | cut -f3 | sort -k3,3 > ${f%.*}.svid 
done
sort -m $s/*.svid | uniq -c | awk '$1>1 {print $2}' > $s/dp.svid

for c in {01..23} 25; do		# chromosomes in UKB imputed data
	echo -n "plink2 --pfile $s/$c --geno .05 --hwe 1e-6 --exclude $s/dp.svid "
	echo    "--make-just-pvar --out $c --memory 4096"
done | hpcwp - -d$d -q1 -t1 -m8 --ln $s --log None
```

The result is  stored in **{s}/ukb/imp/002**, with only the  list of variants in
\*.pvar format.


## Samples

With the remaining variants, filter samples according to these criteria:

  * known genotyping outliers reported by UKB.
    - UKB variable __[22027][]__: Outliers for heterozygosity or missing rate
    - saved as __{p}/dat/phe/gno/out.txt__;
	- see [phe.Rmd][] and [phe.html][]  for more details.
  * callrate < 95% for any of the autosomes.
	
Be sure variant QC have been done, so more samples may pass.
```{sh, eval=FALSE}
cd $GRP/ukb/dat/imp				# working directory
s=001;							# outcome of QC
d=002; mkdir -p $d				# non-duplicate

# gather genotyping outliers
paste $p/dat/phe/{eid,gno/out}.txt | awk 'NR>1 && $2>0 {print $1}' > $d/outlier.id

# samples failing QC for each autosomes
for c in {01..22}; do
	echo $c
	plink2 --pfile $s/$c --extract $d/$c.pvar --mind .05 --make-just-psam --out $d/$c > /dev/null
done

# samples to be excluded
cat $d/*.mindrem.id $d/out.id | sort -u > $d/out.id

# remove low quality samples
for c in {01..23} 25; do
	echo "plink2 --pfile $s/$c --extract $c.pvar --remove out.id --make-pgen --out $c --memory 4096"
done | hpcwp - -d$d -q1 -t2 -m4.5 --log none --ln $s
```
The results are saved to **{p}/dat/imp/002**, PLIN2 PGEN file sets.

-------------------------------------------------------------------------------

[phe.Rmd]: https://github.com/xiaoran831213/ukb/blob/master/src/phe.Rmd "Extracting phenotypes, (Rmarkdown document)"
[phe.html]: https://htmlpreview.github.io/?https://github.com/xiaoran831213/ukb/blob/master/doc/phe.html "Extracting phenotypes (html)"
[22027]:http://biobank.ndph.ox.ac.uk/showcase/field.cgi?id=22027

## Merge variants

Most of the time it is easier and faster to work with separated chromosomes, but
a  merged table  of  variants in  PLINK  BIM  format is  still  useful for  some
analysis:

```{sh, eval=FALSE}
cd $GRP/ukb/dat/imp/002
for f in *.pgen; do plink2 --pfile ${f%.*} --make-just-bim --out ${f%.*}; done
cat *.bim > ../002.bim
```

# Sample Selection

For samples alreay met quality control criteria:

  * for related individuals, retain only one for each kinship.
	- __ur1.txt__ and __ur2.txt__ under __{p}/dat/phe/gno__ masks two
      partitions free from kinship; use __ur1__ for larger sample;
    - derived from sparse kinship matrix, calculated from GWAS calls
      using KINGSOFT and provided by UKB.
  * retain European samples only
    - marked by **eur.txt** under __{p}/dat/phe__;
	- derived from UKB variables;
	- see "demographics" in __phe.html__ or __phe.Rmd__.

Start with post QC genotype in __"{s}/ukb/imp/002"__, save the subset
of samples as __003__.

The script is here:
```{sh, eval=FALSE}
p=$GRP/ukb; cd $p				# project directory
w=dat/imp;  cd $w				# working directory
s=002							# imputed GT past QC
d=003; mkdir -p $d				# unrelated Europeans

# individuals: unrelated, no genotyping outliar, European ancestry only.
paste $p/dat/phe/{eid.col,{dmg/eur,gno/{ur1,out}}.txt} |
    awk -v FS=$'\t' 'NR>1 && $2==1 && $3==1 && $4!=1 {print $1,$1}'> $d/kpp

# qualified individuals
for c in $s/*.bed; do
    c=${c%%.*}
    f=${c##*/}
    echo "plink2 --pfile $c --keep kpp --make-bed --out $f --memory 4096"
done | hpcwp - -d$d -q1 -t1 -m8 --log None
```
